theano/tensor/opt.py

```diff
@@ -3987,11 +3987,31 @@ def local_useless_reshape(node):
     single dimension.
 
     """
-    if isinstance(node.op, T.Reshape):
-        if (node.inputs[0].ndim == 1 and node.outputs[0].ndim == 1 and
-                node.inputs[0].broadcastable ==
-                node.outputs[0].broadcastable):
-            return [node.inputs[0]]
+    op = node.op
+    if not isinstance(op, Reshape):
+        return False
+
+    input = node.inputs[0]
+    output = node.outputs[0]
+    output_shape = node.inputs[1]
+
+    if (input.ndim == 1 and output.ndim == 1 and
+            input.broadcastable == output.broadcastable):
+        return [input]
+
+    dimshuffle_new_order = []
+    new_output_shape = []
+    i = 0  # index over the output of the new reshape
+    for dim in output_shape.value:
+        if dim == 1:
+            dimshuffle_new_order.append('x')
+        else:
+            dimshuffle_new_order.append(i)
+            new_output_shape.append(dim)
+            i = i + 1
+    if len(dimshuffle_new_order) > 0:
+        inner = op.__class__(len(new_output_shape))(input, new_output_shape)
+        return [DimShuffle(inner.type.broadcastable, dimshuffle_new_order)(inner)]
```

Comment:
line 4012
```python
if len(dimshuffle_new_order) > 0:
```
checks for this. If ```len(dimshuffle_new_order)``` is zero then it means that no reshape of 1 is found.